tree/smt/verifier.go

```diff
@@ -5,50 +5,179 @@ import (
+// `oldKey‖oldValue` is the neighbour leaf that prevents the key from being
+// inserted (see the Iden3 SMT paper, §4.3).
+// `isOld0` ≙ 1 when the tree is empty on that branch.
+//
+// It returns 1 on a correct non-membership proof, 0 otherwise.
+func ExclusionVerifier(
+	api frontend.API,
+	hFn utils.Hasher,
+	root frontend.Variable,
+	siblings []frontend.Variable,
+	oldKey, oldValue, isOld0 frontend.Variable, // witness for the existing leaf
+	key frontend.Variable, // the key we claim is absent
+) frontend.Variable {
+	return Verifier(
+		api, hFn, 1, root, siblings,
+		oldKey, oldValue, isOld0,
+		key, 0, 1,
+	)
 }
 
-func Verifier(api frontend.API, hFn utils.Hasher, enabled, root frontend.Variable, siblings []frontend.Variable, oldKey, oldValue, isOld0, key, value, fnc frontend.Variable) {
-	hash1Old := Hash1(api, hFn, oldKey, oldValue)
-	hash1New := Hash1(api, hFn, key, value)
-	VerifierWithLeafHash(api, hFn, enabled, root, siblings, oldKey, hash1Old, isOld0, key, hash1New, fnc)
+// Verifier is the common entry-point for both inclusion and exclusion proofs.
+//
+// * **enabled** – set to 0 to bypass all checks (flag will always be 1).
+// * **fnc**     – 0 ⇒ inclusion, 1 ⇒ exclusion (Iden3 notation).
+//
+// It delegates to the “leaf-hash” version after computing the
+// Poseidon-based leaf commitments (`Hash1`).
+func Verifier(
+	api frontend.API,
+	hFn utils.Hasher,
+	enabled frontend.Variable, // 0 ⇒ proof disabled
+	root frontend.Variable, // public input
+	siblings []frontend.Variable, // packed top→bottom
+	oldKey, oldValue, isOld0 frontend.Variable,
+	key, value frontend.Variable,
+	fnc frontend.Variable, // 0 = inc, 1 = exc
+) frontend.Variable {
+
+	hash1Old := Hash1(api, hFn, oldKey, oldValue) // H(oldKey‖oldValue‖1)
+	hash1New := Hash1(api, hFn, key, value)       // H(key‖value‖1)
+
+	return VerifierWithLeafHashFlag(
+		api, hFn,
+		enabled, root, siblings,
+		oldKey, hash1Old, isOld0,
+		key, hash1New, fnc,
+	)
+}
+
+// VerifierWithLeafHash is a convenience wrapper around VerifierWithLeafHashFlag
+// that asserts the result to be 1 (valid).
+func VerifierWithLeafHash(
+	api frontend.API,
+	hFn utils.Hasher,
+	enabled frontend.Variable,
+	root frontend.Variable,
+	siblings []frontend.Variable,
+	oldKey, hash1Old, isOld0 frontend.Variable,
+	key, hash1New frontend.Variable,
+	fnc frontend.Variable, // 0=inclusion, 1=exclusion
+) {
+	valid := VerifierWithLeafHashFlag(
+		api, hFn,
+		enabled, root, siblings,
+		oldKey, hash1Old, isOld0,
+		key, hash1New, fnc,
+	)
+	api.AssertIsEqual(valid, 1)
 }
 
-func VerifierWithLeafHash(api frontend.API, hFn utils.Hasher, enabled, root frontend.Variable, siblings []frontend.Variable, oldKey, hash1Old, isOld0, key, hash1New, fnc frontend.Variable) {
+// VerifierWithLeafHashFlag rebuilds the root and returns 1 on success, 0 on failure.
+func VerifierWithLeafHashFlag(
+	api frontend.API,
+	hFn utils.Hasher,
+	enabled, // 1 → perform all checks, 0 → bypass (flag always 1)
```

Comment:
Include this comments in the main one of the function header to make them readable when the spec is consumed.